linkerd/proxy/discover/src/from_resolve.rs

```diff
@@ -111,50 +113,70 @@ where
                         }
-                        *this.active = active;
                     }
-                    Update::Add(endpoints) => {
-                        for (addr, endpoint) in endpoints.into_iter() {
+
+                    for (addr, endpoint) in new_active.iter() {
+                        if this.active.get(addr) != Some(endpoint) {
                             trace!(%addr, "Scheduling addition");
-                            this.active.insert(addr);
-                            this.pending.push_back(Change::Insert(addr, endpoint));
+                            this.pending
+                                .push_back(Change::Insert(*addr, endpoint.clone()));
                         }
                     }
-                    Update::Remove(addrs) => {
-                        for addr in addrs.into_iter() {
-                            if this.active.remove(&addr) {
-                                trace!(%addr, "Scheduling removal");
-                                this.pending.push_back(Change::Remove(addr));
+
+                    *this.active = new_active;
+                }
+
+                Update::Add(endpoints) => {
+                    for (addr, endpoint) in endpoints.into_iter() {
+                        trace!(%addr, "Scheduling addition");
+                        match this.active.entry(addr) {
+                            indexmap::map::Entry::Vacant(entry) => {
+                                entry.insert(endpoint.clone());
+                                this.pending.push_back(Change::Insert(addr, endpoint));
+                            }
+                            indexmap::map::Entry::Occupied(mut entry) => {
+                                if entry.get() != &endpoint {
+                                    entry.insert(endpoint.clone());
+                                    this.pending.push_back(Change::Insert(addr, endpoint));
+                                }
                             }
                         }
                     }
-                    Update::DoesNotExist => {
-                        trace!("Scheduling removals");
-                        this.pending.extend(this.active.drain().map(Change::Remove));
+                }
+
+                Update::Remove(addrs) => {
+                    for addr in addrs.into_iter() {
+                        if this.active.remove(&addr).is_some() {
+                            trace!(%addr, "Scheduling removal");
+                            this.pending.push_back(Change::Remove(addr));
+                        }
                     }
-                },
-                None => return Poll::Ready(None),
+                }
+
+                Update::DoesNotExist => {
+                    trace!("Scheduling removals");
+                    this.pending
+                        .extend(this.active.drain(..).map(|(sa, _)| Change::Remove(sa)));
+                    this.active.clear();
```

Comment:
i don't *think* the `clear` call should be necessary here, since the `drain(..)` iterator should remove everything from `this.pending`? if this actually is load-bearing, i would kind of like a comment explaining why, otherwise, we could just remove it.